foundation-models/openai/src/main/java/com/sap/ai/sdk/foundationmodels/openai/OpenAiTool.java

```diff
@@ -0,0 +1,192 @@
+import com.github.victools.jsonschema.generator.SchemaGenerator;
+import com.github.victools.jsonschema.generator.SchemaGeneratorConfigBuilder;
+import com.github.victools.jsonschema.generator.SchemaVersion;
+import com.github.victools.jsonschema.module.jackson.JacksonModule;
+import com.github.victools.jsonschema.module.jackson.JacksonOption;
+import com.google.common.annotations.Beta;
+import com.sap.ai.sdk.foundationmodels.openai.generated.model.ChatCompletionTool;
+import com.sap.ai.sdk.foundationmodels.openai.generated.model.FunctionObject;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.experimental.Accessors;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Represents an OpenAI tool that can be used to define a function call in an OpenAI Chat Completion
+ * request. This tool generates a JSON schema based on the provided class representing the
+ * function's request structure.
+ *
+ * @param <InputT> the type of the input argument for the function
+ * @see <a href="https://platform.openai.com/docs/guides/gpt/function-calling"/>OpenAI Function
+ * @since 1.7.0
+ */
+@Slf4j
+@Beta
+@Data
+@Getter(AccessLevel.PACKAGE)
+@Accessors(chain = true)
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+public class OpenAiTool<InputT> {
+
+  private static final ObjectMapper JACKSON = new ObjectMapper();
+
+  /** The schema generator used to create JSON schemas. */
+  @Nonnull private static final SchemaGenerator GENERATOR = createSchemaGenerator();
+
+  /** The name of the function. */
+  @Nonnull private String name;
+
+  /** The model class for function request. */
+  @Nonnull private Class<InputT> requestClass;
+
+  /** An optional description of the function. */
+  @Nullable private String description;
+
+  /** An optional flag indicating whether the function parameters should be treated strictly. */
+  @Nullable private Boolean strict;
+
+  /** The function to be called. */
+  @Nullable private Function<InputT, ?> function;
+
+  /**
+   * Constructs an {@code OpenAiFunctionTool} with the specified name and a model class that
+   * captures the request to the function.
+   *
+   * @param name the name of the function
+   * @param requestClass the model class for function request
+   */
+  public OpenAiTool(@Nonnull final String name, @Nonnull final Class<InputT> requestClass) {
+    this(name, requestClass, null, null, null);
+  }
+
+  @Nonnull
+  Object execute(@Nonnull final InputT argument) {
+    if (getFunction() == null) {
+      throw new IllegalStateException(
+          "Tool " + name + " is missing a method reference to execute.");
```

Comment:
```suggestion
          "Please set the function for tool: " + name);
```
Give an actionable error for the user

Reply:
Interesting point. What's better for an exception message: _inherent cause_ or _missing action_ 🤔
Anyway, I agree the message is not optimal

```suggestion
          "Tool is not executable. Please set a function for tool: " + name);
```